File: code/aiming_transporting/matlab/modules/aiming.m
function [theta_val, tau_val, phi_val] = solve_theta(params)
% 求解电子束发射角度和飞行参数

% 定义辅助函数
a0 = @(theta) params.a0_max * sin(theta);

theta_equation = @(theta) params.R_m * sin(params.Psi) - ...
    a0(theta) * sqrt(2 * (1 - cos(params.wB * params.R_m * cos(params.Psi) / ...
    (params.v0 * cos(theta))));

% 寻找解区间
theta_min = 0.01;
theta_max = pi/2;
theta_test = linspace(theta_min, theta_max, 1000);
f_test = arrayfun(theta_equation, theta_test);

bracket = [];
for i = 1:length(f_test)-1
    if f_test(i) * f_test(i+1) < 0
        bracket = [theta_test(i), theta_test(i+1)];
        break;
    end
end

if ~isempty(bracket)
    options = optimset('Display','off');
    [theta_val, ~, exitflag] = fzero(theta_equation, bracket, options);
    
    if exitflag ~= 1
        error('未解出合适打击轨迹，请检查输入参数');
    end
else
    error('未找到合适的theta区间');
end

% 计算其他参数
tau_val = params.R_m * cos(params.Psi) / params.v0 / cos(theta_val);
a0_val = a0(theta_val);
xT = a0_val * (1 - cos(params.wB * tau_val));
yT = a0_val * sin(params.wB * tau_val);
phi_val = atan2(-xT, yT);
end

function check_range(params)
% 检查目标是否在射程内
max_range = 2 * params.a0_max / sin(params.Psi);
if params.R_m > max_range
    error('目标距离超出射程！当前目标角度下射程需小于: %.1f公里', max_range*1e-3);
end
end

File: code/aiming_transporting/modules/aiming.py
from modules.beam_params import BeamParams
import numpy as np
from scipy.optimize import root_scalar

# 判断是否超出射程
def check_range(params: BeamParams):
    if params.R_m > 2 * params.a0_max / np.sin(params.Psi):
        raise ValueError(f"目标距离超出射程！当前目标角度下射程需小于: {2*params.a0_max/np.sin(params.Psi)*1e-3:.1f}公里，请选择距离或角度更小的目标后重试")

# 求解发射角theta 
def solve_theta(params: BeamParams):
    def a0(theta):
        return params.a0_max * np.sin(theta)

    def theta_equation(theta):
        a0_val = a0(theta)
        return params.R_m * np.sin(params.Psi) - a0_val * np.sqrt(2 * (1 - np.cos(params.wB * params.R_m * np.cos(params.Psi) / (params.v0 * np.cos(theta)))))

    theta_min, theta_max = 0.01, np.pi/2
    theta_test = np.linspace(theta_min, theta_max, 1000)
    f_test = [theta_equation(theta) for theta in theta_test]
    bracket = None
    for i in range(len(f_test)-1):
        if f_test[i] * f_test[i+1] < 0:
            bracket = [theta_test[i], theta_test[i+1]]
            break
    if bracket is not None:
        sol = root_scalar(lambda theta: theta_equation(theta), bracket=bracket, method='brentq')
        theta_val = sol.root if sol.converged else None
        if theta_val is None:
            raise ValueError("未解出合适打击轨迹，请检查输入参数")
    else:
        raise ValueError("未找到合适的theta区间")
    tau_val = params.R_m * np.cos(params.Psi) / params.v0 / np.cos(theta_val)
    a0_val = params.a0_max * np.sin(theta_val)
    xT = a0_val * (1 - np.cos(params.wB * tau_val))
    yT = a0_val * np.sin(params.wB * tau_val)
    phi_val = np.arctan2(-xT, yT)
    return theta_val, tau_val, phi_val
